cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java

```diff
@@ -0,0 +1,178 @@
+
+/**
+ * Metadata Migrator to migrate metadata from V1 metadata tables to V2 metadata tables.
+ * Migration happens on business table as well as system table. It only migrates value and history rows.
+ * Indexes will be automatically generated for every write on V2 metadata table.
+ */
+class MetadataMigrator extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(MetadataMigrator.class);
+  // For outage, only log once per 60 seconds per message.
+  private static final Logger OUTAGE_LOG = Loggers.sampling(LOG, LogSamplers.perMessage(
+    () -> LogSamplers.limitRate(60000)));
+  private static final List<DatasetId> DATASET_IDS = ImmutableList.of(NamespaceId.SYSTEM.dataset("system.metadata"),
+                                                                      NamespaceId.SYSTEM.dataset("v2.system.metadata"),
+                                                                      NamespaceId.SYSTEM.dataset("business.metadata"),
+                                                                      NamespaceId.SYSTEM.dataset("v2.business.metadata")
+  );
+
+  private final DatasetFramework dsFramework;
+  private final Transactional transactional;
+  private final int batchSize;
+  private volatile Thread runThread;
+  private volatile boolean stopped;
+  private volatile boolean hasV1Instance;
+
+  MetadataMigrator(CConfiguration cConf, DatasetFramework dsFramework, TransactionSystemClient txClient) {
+    this.dsFramework = dsFramework;
+    this.batchSize = cConf.getInt(Constants.Metadata.MIGRATOR_BATCH_SIZE);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(new SystemDatasetInstantiator(dsFramework),
+                                                                   txClient, NamespaceId.SYSTEM,
+                                                                   Collections.emptyMap(), null, null)),
+      org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting Metadata Migrator Service.");
+  }
+
+  @Override
+  public void run() {
+    runThread = Thread.currentThread();
+
+    try {
+      for (int i = 0; i < DATASET_IDS.size(); i = i + 2) {
+        // assume we already have v1 metadata table instance
+        hasV1Instance = true;
+
+        while (!stopped && hasV1Instance) {
+          try {
+            final int index = i;
+            // This thread migrates metadata in batches. It does following operations in a single transaction:
+            // 1.) Scan V1 metadata table for value and history rows in batch of batchSize.
+            // 2.) Write scanned MetadataEntries to V2 table.
+            // 3.) Delete successfully written metadata entries from V1 table.
+            Transactionals.execute(transactional, context -> {
+              if (!dsFramework.hasInstance(DATASET_IDS.get(index))) {
```

Comment:
This check can be outside of the `Transactionals.execute`, and I think it can even be outside of the while loop. In fact, the `hasV1Instance` can be a local variable.

Reply:
+1 for calling `hasInstance` outside transaction. But keeping this in while loop will help catch the exception and retry.